File: implementations/python/pairing_heap.py
"""
Reference Implementation: Self-Adjusting Pairing Heap
Demonstrates the two-pass pairing algorithm (Fredman et al., 1986),
O(1) meld, O(1) insert, O(log n) amortized delete_min,
and first-child / next-sibling representation.

Language: Python 3
"""

from typing import Any, Optional, List, Callable
import unittest


class PairingNode:
    def __init__(self, val: Any):
        self.val: Any = val
        self.child: Optional["PairingNode"] = None
        self.sibling: Optional["PairingNode"] = None


class PairingHeap:
    def __init__(self, key: Optional[Callable[[Any], Any]] = None):
        self._root: Optional[PairingNode] = None
        self._size: int = 0
        self._key: Callable[[Any], Any] = key if key is not None else (lambda x: x)

    def __len__(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._root is None

    def _meld_nodes(self, a: Optional[PairingNode], b: Optional[PairingNode]) -> Optional[PairingNode]:
        if not a:
            return b
        if not b:
            return a

        if self._key(b.val) < self._key(a.val):
            a, b = b, a

        # 'b' becomes the leftmost child of 'a'
        b.sibling = a.child
        a.child = b
        return a

    def _two_pass_merge(self, first: Optional[PairingNode]) -> Optional[PairingNode]:
        if not first:
            return None
        if not first.sibling:
            return first

        # Pass 1: Left-to-Right pairing
        pairs = []
        curr = first
        while curr:
            a = curr
            b = curr.sibling
            if b:
                next_node = b.sibling
                a.sibling = None
                b.sibling = None
                pairs.append(self._meld_nodes(a, b))
                curr = next_node
            else:
                a.sibling = None
                pairs.append(a)
                curr = None

        # Pass 2: Right-to-Left accumulation
        result = pairs[-1]
        for i in range(len(pairs) - 2, -1, -1):
            result = self._meld_nodes(pairs[i], result)

        return result

    def peek(self) -> Any:
        if not self._root:
            raise IndexError("peek from empty heap")
        return self._root.val

    def push(self, val: Any) -> None:
        new_node = PairingNode(val)
        self._root = self._meld_nodes(self._root, new_node)
        self._size += 1

    def pop(self) -> Any:
        if not self._root:
            raise IndexError("pop from empty heap")
        top_val = self._root.val
        children = self._root.child
        self._root = self._two_pass_merge(children)
        self._size -= 1
        return top_val

    def meld(self, other: "PairingHeap") -> None:
        if self is other or other.is_empty():
            return
        self._root = self._meld_nodes(self._root, other._root)
        self._size += other._size
        other._root = None
        other._size = 0


class TestPairingHeap(unittest.TestCase):
    def test_basic_operations(self):
        heap = PairingHeap()
        self.assertTrue(heap.is_empty())
        self.assertEqual(len(heap), 0)

        inputs = [42, 17, 93, 8, 31, 5, 64, 22, 11, 75]
        for x in inputs:
            heap.push(x)

        self.assertEqual(len(heap), len(inputs))
        self.assertEqual(heap.peek(), 5)

        sorted_out = []
        while not heap.is_empty():
            sorted_out.append(heap.pop())

        self.assertEqual(sorted_out, sorted(inputs))
        self.assertTrue(heap.is_empty())

    def test_meld(self):
        h1 = PairingHeap()
        for x in [10, 30, 50]:
            h1.push(x)

        h2 = PairingHeap()
        for x in [5, 25, 70]:
            h2.push(x)

        h1.meld(h2)
        self.assertEqual(len(h1), 6)
        self.assertTrue(h2.is_empty())
        self.assertEqual(h1.peek(), 5)

        sorted_out = []
        while not h1.is_empty():
            sorted_out.append(h1.pop())

        self.assertEqual(sorted_out, [5, 10, 25, 30, 50, 70])


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/pairing_heap.cpp
/**
 * Reference Implementation: Self-Adjusting Pairing Heap
 * Demonstrates the two-pass pairing algorithm (Fredman et al., 1986),
 * O(1) meld, O(1) insert, O(log n) amortized delete_min,
 * and first-child / next-sibling representation.
 *
 * Language: C++17
 */

#include <iostream>
#include <vector>
#include <functional>
#include <cassert>
#include <stdexcept>
#include <utility>
#include <algorithm>

template <typename T, typename Compare = std::less<T>>
class PairingHeap {
public:
    struct Node {
        T val;
        Node* child;
        Node* sibling;

        explicit Node(T v) : val(std::move(v)), child(nullptr), sibling(nullptr) {}
    };

private:
    Node* root_;
    std::size_t size_;
    Compare comp_; // Returns true if a has higher priority than b

    static Node* meld_nodes(Node* a, Node* b, Compare comp) {
        if (!a) return b;
        if (!b) return a;

        // Ensure 'a' has the higher priority
        if (comp(b->val, a->val)) {
            std::swap(a, b);
        }

        // 'b' becomes the leftmost child of 'a'
        b->sibling = a->child;
        a->child = b;
        return a;
    }

    // Canonical Two-Pass Pairing Algorithm (Fredman, Sedgewick, Sleator, Tarjan 1986)
    static Node* two_pass_merge(Node* first, Compare comp) {
        if (!first) return nullptr;
        if (!first->sibling) return first;

        // Pass 1: Left-to-Right pairing
        std::vector<Node*> pairs;
        Node* curr = first;
        while (curr) {
            Node* a = curr;
            Node* b = curr->sibling;
            if (b) {
                Node* next = b->sibling;
                a->sibling = nullptr;
                b->sibling = nullptr;
                pairs.push_back(meld_nodes(a, b, comp));
                curr = next;
            } else {
                a->sibling = nullptr;
                pairs.push_back(a);
                curr = nullptr;
            }
        }

        // Pass 2: Right-to-Left accumulation
        Node* result = pairs.back();
        for (int i = static_cast<int>(pairs.size()) - 2; i >= 0; --i) {
            result = meld_nodes(pairs[i], result, comp);
        }
        return result;
    }

    static void destroy_tree(Node* n) noexcept {
        // Non-recursive destruction to prevent stack overflow on deep trees
        std::vector<Node*> stack;
        if (n) stack.push_back(n);

        while (!stack.empty()) {
            Node* curr = stack.back();
            stack.pop_back();

            if (curr->sibling) stack.push_back(curr->sibling);
            if (curr->child) stack.push_back(curr->child);
            delete curr;
        }
    }

public:
    explicit PairingHeap(Compare comp = Compare{})
        : root_(nullptr), size_(0), comp_(comp) {}

    ~PairingHeap() {
        destroy_tree(root_);
    }

    PairingHeap(const PairingHeap&) = delete;
    PairingHeap& operator=(const PairingHeap&) = delete;

    PairingHeap(PairingHeap&& other) noexcept
        : root_(other.root_), size_(other.size_), comp_(other.comp_) {
        other.root_ = nullptr;
        other.size_ = 0;
    }

    PairingHeap& operator=(PairingHeap&& other) noexcept {
        if (this != &other) {
            destroy_tree(root_);
            root_ = other.root_;
            size_ = other.size_;
            comp_ = other.comp_;
            other.root_ = nullptr;
            other.size_ = 0;
        }
        return *this;
    }

    [[nodiscard]] std::size_t size() const noexcept { return size_; }
    [[nodiscard]] bool empty() const noexcept { return size_ == 0; }

    const T& top() const {
        if (!root_) {
            throw std::underflow_error("PairingHeap::top(): heap is empty");
        }
        return root_->val;
    }

    void push(const T& val) {
        Node* node = new Node(val);
        root_ = meld_nodes(root_, node, comp_);
        ++size_;
    }

    void push(T&& val) {
        Node* node = new Node(std::move(val));
        root_ = meld_nodes(root_, node, comp_);
        ++size_;
    }

    T pop() {
        if (!root_) {
            throw std::underflow_error("PairingHeap::pop(): heap is empty");
        }
        T top_val = std::move(root_->val);
        Node* old_root = root_;
        Node* children = root_->child;

        delete old_root;
        root_ = two_pass_merge(children, comp_);
        --size_;
        return top_val;
    }

    // Melds another PairingHeap into this one in O(1) time
    void meld(PairingHeap& other) {
        if (this == &other || other.empty()) return;

        root_ = meld_nodes(root_, other.root_, comp_);
        size_ += other.size_;

        other.root_ = nullptr;
        other.size_ = 0;
    }
};

int main() {
    PairingHeap<int> min_heap;
    assert(min_heap.empty());
    assert(min_heap.size() == 0);

    // 1. Push elements
    std::vector<int> inputs = {42, 17, 93, 8, 31, 5, 64, 22, 11, 75};
    for (int x : inputs) {
        min_heap.push(x);
    }
    assert(min_heap.size() == inputs.size());
    assert(min_heap.top() == 5);

    // 2. Pop all and verify sorted order
    std::vector<int> sorted;
    while (!min_heap.empty()) {
        sorted.push_back(min_heap.pop());
    }
    assert(sorted == std::vector<int>({5, 8, 11, 17, 22, 31, 42, 64, 75, 93}));
    assert(min_heap.empty());

    // 3. O(1) Meld verification
    PairingHeap<int> h1;
    h1.push(10);
    h1.push(30);
    h1.push(50);

    PairingHeap<int> h2;
    h2.push(5);
    h2.push(25);
    h2.push(70);

    h1.meld(h2);
    assert(h1.size() == 6);
    assert(h2.empty());
    assert(h1.top() == 5);

    std::vector<int> melded_sorted;
    while (!h1.empty()) {
        melded_sorted.push_back(h1.pop());
    }
    assert(melded_sorted == std::vector<int>({5, 10, 25, 30, 50, 70}));

    // 4. Max-Heap comparator verification
    PairingHeap<int, std::greater<int>> max_heap;
    for (int x : inputs) {
        max_heap.push(x);
    }
    assert(max_heap.top() == 93);
    assert(max_heap.pop() == 93);
    assert(max_heap.pop() == 75);

    std::cout << "[PASS] All PairingHeap C++ unit tests passed.\n";
    return 0;
}
